src/Bounds.cpp

```diff
@@ -1046,6 +1047,42 @@ bool box_contains(const Box &outer, const Box &inner) {
     return is_one(simplify(condition));
 }
 
+// Place innermost vars in an IfThenElse's condition as far to the left as possible.
+class SolveIfThenElse : public IRMutator {
+    // List of variables (LetStmt or For) encountered starting from outermost
+    // to innermost
+    vector<string> vars;
+
+    using IRMutator::visit;
+
+    void visit(const LetStmt *op) {
+        vars.push_back(op->name);
+        IRMutator::visit(op);
+        vars.pop_back();
+    }
+
+    void visit(const For *op) {
+        vars.push_back(op->name);
+        IRMutator::visit(op);
+        vars.pop_back();
+    }
+
+    void visit(const IfThenElse *op) {
+        IRMutator::visit(op);
+        op = stmt.as<IfThenElse>();
+        internal_assert(op);
+        for (size_t i = vars.size(); i > 0; --i) {
```

Comment:
What is the performance cost of this? I worry about specialization, which creates ifthenelse nodes based on params. If there are lots of variables in scope, this will iterate over all of them, checking if each is used in the expression. An alternative is to maintain a map<string, int> which gives the nesting depth of each variable, then write a visitor that finds the variable in an expr with maximum nesting depth, and then solve for that variable.